### ISSUE TITLE
Kubernetes api, controller, and client performance work

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#27667]
This is a roundup issue for Kubernetes 1.4 and a tracker to orient folks on the ongoing work.  It is focused on all components that touch the API, but not necessarily the total system.  It also serves to identify areas where folks can get involved.

Track issues using [labels performance and CSI-API-Machinery](https://github.com/kubernetes/kubernetes/issues?utf8=✓&q=label%3Aarea%2Fperformance+label%3A%22team%2FCSI-API+Machinery+SIG%22+)
### Ongoing efforts

Major focus: We want to get to our target cluster size metrics (5k nodes and 300k pods) in 1.4 or 1.5, and we are within an order of magnitude of that with 1.3 in many dimensions.

Areas of attack: We are mostly CPU bound on the controllers and API servers today, and most of that is overhead due to garbage collection.  We also have had a fair number of contention areas that were algorithmic or lock related in the core components.  We've got a good feedback loop for identifying critical bottlenecks, but we could probably stand to get more profiles and metrics gathered from the high scale systems.
- Reduce the CPU and garbage created by clients and the API server when handling API requests
  - Introduce protobuf
    - [complete #8132] Used for most API objects when talking between client and server
    - [inprogress for 1.5] Use protobuf for storage format
    - Handle mixed content type storage (objects that embed other objects like Templates or third party resources)
    - [inprogress #27258] Handle optional slices and maps correctly in protobuf
    - [inprogress](https://github.com/gogo/protobuf/issues/174) Investigate use of string arena unmarshalling in protobuf to avoid []byte -> string allocations in decode
  - Optimize the conversion, defaulting, and decoding code paths
    - [complete](https://github.com/kubernetes/kubernetes/issues/25019) Implement more efficient Quantity, using int64 and better serialization
    - [complete #24966] Remove the use of reflection in conversion and defaulting
      - Auto generate fast path conversion functions (AddGenericConversionFunc) for each API type
    - [complete #24727] Make conversions of identical types (builtins, maps, slices, pointers) an assignment, not a copy
    - [inprogress #27659] Further optimization of identical type conversion using unsafe.Pointer
    - [inprogress #33352] Split defaulting from conversion into its own function, possibly by generating an entire "Default_Pod" function from component parts, which allows us to avoid having to walk deep objects
  - Use the versioned client where possible to avoid conversion altogether. https://github.com/kubernetes/kubernetes/issues/29934
    - [inprogress](???) Versioned client in controllers depends on the shared cache work (all controllers sharing a cache have to use the same object type). It's also blocked by the [cutting a client library work](https://github.com/kubernetes/kubernetes/pull/29147).
  - Ensure that the Update and Create code paths are efficient (guaranteed update loop generates a lot of objects)
  - SelfLink and URL.String() are fairly expensive in the encode/decode path
  - For any JSON gaps (non-proto clients) we could look at easyjson, which appears to do more than ugorgji to reduce memory allocation during decode.
- Reduce the amount of duplicate objects being used by controller and admission caches and the amount of total memory used by caches
  - [complete] Separate re-listing and re-syncing (former talks to api server, latter traverses cache) to reduce the amount of duplicate calls to the server and wasted conversion
  - [complete](https://github.com/kubernetes/kubernetes/issues/14978) Add a SharedInformer and use it for Pod caches (all 8 of them)
  - [inprogress] Use shared Namespace caches in admission controllers
  - Investigate and identify other opportunities for shared caches
  - [inprogress] Investigate size of objects kept in shared and unshared caches for duplication and waste
    - Pods tend to have larger objects https://github.com/coreos/kscale/pull/37/files#diff-a423e914cbd27c95d2c0b950e2e4ee5bR14 than necessary - maps and quantity too large
    - Investigate using slices instead of maps for our many small maps, or having an `ArrayMap` equivalent to Java [#31440](https://github.com/kubernetes/kubernetes/issues/31440)
    - Investigate flyweight cache patterns for certain controllers
- Identify unnecessary blocking / stalls in clients, watches, caches, and controllers
  - [complete] Several fixes to 1.3 scheduler and caches
  - [complete] Avoid flushing aggressively in watches to reduces syscalls
  - [inprogress](@deads2k) Shared caches currently have coarse locks, investigate striping and reducing the duration of the locks
  - Investigate whether we have any slow cache reader problems with our shared caches
  - Audit the watch cache for locking issues
  - Better tooling and documentation to allow people to perform this sort of debugging
  - ???
- Profile and optimize specific components and workloads 
  - Collect profile traces from kubemark and high scale clusters automatically
  - [Scheduler](https://github.com/kubernetes/kubernetes/issues/18418) (lots of eyes on this, but also lots of algorithmic improvements possible)
  - Kubelet
  - [inprogress] Replication controller
  - Node controller (known issues with Timer, other garbage creation)
  - Quota 
    - [inprogress] Significant improvements including contention striping and batch assignment
    - Optimize and audit use of resource.Quantity in these codepaths
  - Authentication / Authorization / RBAC (has been a significant source of perf bottlenecks in OpenShift in the more complex scenarios)
  - ???
- Reduce traffic between components
  - Investigation colocation of certain processes if necessary
    - OpenShift offers all-in-one processes that bundle etcd, controller, apiserver, as does a few others, and there is some value in compact deployments having an efficient master (downsides are increased memory use and contention)
    - [inprogress](https://github.com/openshift/origin/pull/8082) Investigate loopback connections when in the same process (at the net.Chan level)
    - [mentioned by justinsb] Directly embed etcd and make calls to the Go API
  - Reduce network bandwidth by sending less objects (not yet a blocker in any particular area)

**Comment by smarterclayton** [comment]
@kubernetes/sig-api-machinery @kubernetes/sig-scalability @xiang90 @hongchaodeng as there have been a lot of people recently discussing what they're doing in 1.4, wanted to make sure we had a list of the areas everyone is attacking.  Cross-over between the two sigs.

**Comment by davidopp** [comment]
We have a separate issue for tracking scheduler performance/scalability improvements: #18418

**Comment by wojtek-t** [comment]
@smarterclayton - thanks for putting this together. I have few other things on my mind, but let me get back to it on Monday.

One thing that gives significant improvements on scale is avoid a lot of unnecessary computations in cacher - in large clusters we know that only O(1) watchers are interested in all pods/nodes and most of them are iterested in only one node/pods from that node.

I have already a PR in-flight https://github.com/kubernetes/kubernetes/pull/27277 that seems to save ~75% of cpu usage in apiserver in empty 2k-node cluster.  This may be enough or we may consider doing it more sophisticated. But for 2k-node clusters this seem to give huge improvements.

I have some other thought, but let me write them on Monday.

**Comment by caesarxuchao** [comment]
> Use the versioned client where possible to avoid conversion altogether
inprogress Versioned client in controllers depends on the shared cache work (all controllers sharing a cache have to use the same object type)

Thanks @smarterclayton. The link you gave doesn't work. Did you mean to link to this [issue](https://github.com/kubernetes/kubernetes/issues/4874)? I have an outdated [WIP PR](https://github.com/kubernetes/kubernetes/pull/24155) related to this and plan to work on it in 1.4.

**Comment by hongchaodeng** [comment]
> most of them are iterested in only one node/pods from that node.

If so, it should watch on single pod, right?

**Comment by wojtek-t** [comment]
It's not a single pod - these are all pods attached to a given node.

**Comment by smarterclayton** [comment]
Closing, has fulfilled purpose.
